### PR TITLE
Better ElasticDeviceMesh

### PR DESCRIPTION
**Pull request by Jackmin801**
- [x] Patch into simulate-multi-node.sh
- [x] Fix sync

### CODE REVIEW

File: src/zeroband/diloco.py [hunk 1]
```diff
@@ -108,29 +94,42 @@ def sync_inner_model(self, model: nn.Module):
...
+                offloaded_param = torch.tensor(storage, dtype=param.dtype, device="cpu")
+                offloaded_param.as_strided_(size=param.data.size(), stride=param.data.stride())
+                if self.world_info.rank == 0:
+                    # TODO: Can we async or split the copy among gpus probs overkill?
+                    offloaded_param.copy_(param.data)
+                offloaded_param.requires_grad = False # TODO: check if we need to set this to True
                 offloaded_params.append(offloaded_param)
 
+        dist.barrier()
         return offloaded_params
```
Comment (Jackmin801) [inline src/zeroband/diloco.py]: The change here makes processes which are part of the same FSDP replica group (which are assumed to be on the same node with the same /dev/shm) use the same underlying storage for the offloaded_param.

All the processes use the same shared memory file to create a storage for each parameter. Only rank 0 will do the copy.

A barrier is added to ensure that after the function completes, the parameters are all offloaded. Otherwise, the non 0 ranks might access uninitialized memory.

File: src/zeroband/diloco.py [hunk 2]
```diff
@@ -108,29 +94,42 @@ def sync_inner_model(self, model: nn.Module):
...
         """
-        self.sync_pseudo_gradient(model)
-        if self.outer_optimizer is not None:
-            self.outer_optimizer.step()
-            self.outer_optimizer.zero_grad()  # todo(sami): check if we can remove this
+        if self.world_info.rank == 0:
+            self.sync_pseudo_gradient(model)
+            if self.outer_optimizer is not None:
+                self.outer_optimizer.step()
+                self.outer_optimizer.zero_grad()  # todo(sami): check if we can remove this
```
Comment (Jackmin801) [inline src/zeroband/diloco.py]: we probably can. we never use backward or add to the grad. they are always set so should be fine to not zero here.

File: scripts/simulate_multi_node_diloco.sh [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+mkdir -p logs
+
+export GLOBAL_ADDR=localhost
+export GLOBAL_PORT=10000
+export GLOBAL_WORLD_SIZE=$N
+
+for i in $(seq 0 $(($N - 1 )))
+do
+    > logs/log$i
+    TORCH_UNIQUE_ID=$i GLOBAL_RANK=$i CUDA_VISIBLE_DEVICES=$(get_cuda_devices $NUM_GPU $i) uv run torchrun --nproc_per_node=$NUM_GPU --node-rank 0 --rdzv-endpoint localhost:$((10001 + $i)) --nnodes=1  $@  > logs/log$i 2>&1 &
```
Comment (Jackmin801) [inline scripts/simulate_multi_node_diloco.sh]: These are the only changes. We need `GLOBAL_*` to initialize the outer process group's store. TORCH_UNIQUE_ID is used to make sure processes see the correct shared parameters and is also used by the on-off ramping logic to communicate the new GLOBAL_RANK to processes.

The `rdvz-endpoint` was changed so that the multiple runs dont interfere with each other. `nnodes=1 and node-rank=0` because the nodes need to initialize their own TCP store for the FSDP.

File: src/zeroband/diloco.py [hunk 4]
```diff
@@ -108,29 +94,42 @@ def sync_inner_model(self, model: nn.Module):
...
 
+        dist.barrier()
```
Comment (Jackmin801) [inline src/zeroband/diloco.py]: the barrier here is so the update is applied to the offload_params before all ranks try to read from the offload_params

File: src/zeroband/testing.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    """
+    Get the module signature
+    """
+    state_dict_sig = {name: get_tensor_signature(param) for name, param in module.named_parameters()}
+    if compress:
+        return hashlib.md5(str(state_dict_sig).encode("utf-8")).hexdigest()
+    else:
+        return '\n'.join(f"{name}: {sig}" for name, sig in state_dict_sig.items())
+
+
```
Comment (Jackmin801) [inline src/zeroband/testing.py]: Used for debugging purposes to make sure the collectives are behaving as expected. We could use them in our tests too maybe.

File: src/zeroband/train.py [hunk 6]
```diff
@@ -128,7 +142,8 @@ def train(config: Config):
         if world_info.local_world_size == 1:
             raise ValueError("Diloco is not supported for local_world_size == 1 because of a pytorch bug")
 
-        diloco = Diloco(config.diloco, model, sharding_strategy, elastic_device_mesh)
+        with FSDP.summon_full_params(model):
+            diloco = Diloco(config.diloco, model, sharding_strategy, elastic_device_mesh)
```
Comment (Jackmin801) [inline src/zeroband/train.py]: Eh we might be able to just move this up and not need the summon

File: src/zeroband/train.py [hunk 7]
```diff
@@ -110,7 +112,19 @@ def train(config: Config):
...
+        for param in model.parameters():
+            # TODO: Kinda ugly but somethings wrong with the world registration
+            #dist.broadcast(param.data, src=0, group=elastic_device_mesh.global_pg)
+            opts = BroadcastOptions()
+            opts.rootRank = 0
+            opts.rootTensor = 0
+            elastic_device_mesh.global_pg.broadcast([param.data], opts)
+
+    for param in model.parameters():
+        dist.broadcast(param.data, src=0, group=elastic_device_mesh.local_pg)
```
Comment (Jackmin801) [inline src/zeroband/train.py]: All these broadcasts are just a result of the non-deterministic initialization. Can be removed once that is fixed. Ill look into how to make the global broadcast less ugly but maybe just leave it for now since it works.

File: src/zeroband/comms.py [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+            status = _wait_for_status(store)
+        
+        if status == "init":
+            # First time initialization
+            self.mesh_count = 0
+            self.prefix_store = dist.PrefixStore("mesh_0", store)
+            pg = dist.ProcessGroupGloo(self.prefix_store, global_rank, global_world_size, TCPSTORE_TIMEOUT)
+            if global_rank == 0:
+                store.set("status", "running")
+            store.set(f"rank_{self.unique_id}", str(global_rank))
```
Comment (Jackmin801) [inline src/zeroband/comms.py]: The "running" path is untested. That is part of the on-off ramping. I think its better to do it in a separate PR because it needs to work with the live checkpoint recovery as well.

File: src/zeroband/comms.py [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+        if self.world_info.rank == 0:
+            self._logger.debug(f"global pg world : {self.global_pg.size()}, local pg: {self.local_pg.size()}")
+        else:
+            self._logger.debug(f"local pg world : {self.local_pg.size()}")
+
+    def __del__(self):
+        dist.destroy_process_group()
+
+    def _init_global_pg(self) -> dist.Store:
+        global_addr = os.environ["GLOBAL_ADDR"]
```
Comment (samsja) [inline src/zeroband/comms.py]: can you move this os.environ stuff into the world info ? Or does it need to be dinamicall ?

File: src/zeroband/comms.py [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+        global_port = int(os.environ["GLOBAL_PORT"])
+        global_world_size = int(os.environ["GLOBAL_WORLD_SIZE"])
+        global_rank = int(os.environ["GLOBAL_RANK"])
```
Comment (samsja) [inline src/zeroband/comms.py]: what the difference between global rank and the other rank ?

File: src/zeroband/comms.py [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+                store.set("status", "running")
+            store.set(f"rank_{self.unique_id}", str(global_rank))
+        elif status == "running":
+            # Node wants to join
+            _queue_join(store, self.unique_id)
+            _wait_for_status(store, "reinit")
+            # Get assigned rank
+            global_rank = int(store.get(f"rank_{self.unique_id}").decode("utf-8"))
+            # Get updated world_size
+            global_world_size = int(store.get("world_size").decode("utf-8"))
```
Comment (samsja) [inline src/zeroband/comms.py]: out of curiosity here what happened if a key is not present in the store, does it crash ?

File: src/zeroband/comms.py [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+        self.leaving = False
+        return pg
+
+    def _init_unique_id(self):
+        """Initialize a unique ID for the node.
+        If TORCH_UNIQUE_ID is set, use that.
+        Otherwise, local rank 0 generates an ID and broadcasts to other nodes.
+        """
+        if "TORCH_UNIQUE_ID" in os.environ:
+            self.unique_id = os.environ["TORCH_UNIQUE_ID"]
```
Comment (samsja) [inline src/zeroband/comms.py]: same ideally put this into world info

File: src/zeroband/comms.py [hunk 13]
```diff
@@ -0,0 +1,261 @@
...
+            return
+        if self.local_rank == 0:
+            self.unique_id = str(uuid.uuid4())
+            with open("/tmp/torch_unique_id", "w") as f:
```
Comment (samsja) [inline src/zeroband/comms.py]: can you add in comments the reason to use file here ?

File: src/zeroband/comms.py [hunk 14]
```diff
@@ -0,0 +1,261 @@
...
+                f.write(self.unique_id)
+        else:
+            while True:
+                try:
+                    with open("/tmp/torch_unique_id", "r") as f:
```
Comment (samsja) [inline src/zeroband/comms.py]: should we rather use `/tmp/zeroband_{some_id}` with some run id or just `/tmp/zeroband` ? I am a bit scared with namespace conflict in shared tmp folder

File: src/zeroband/train.py [hunk 15]
```diff
@@ -116,7 +118,20 @@ def train(config: Config):
...
+    if world_info.rank == 0:
+        for param in model.parameters():
+            # TODO: Kinda ugly but somethings wrong with the world registration
+            # dist.broadcast(param.data, src=0, group=elastic_device_mesh.global_pg)
+            opts = BroadcastOptions()
+            opts.rootRank = 0
+            opts.rootTensor = 0
+            elastic_device_mesh.global_pg.broadcast([param.data], opts)
+
+    for param in model.parameters():
```
Comment (samsja) [inline src/zeroband/train.py]: was it done for weight init ? If yes lets remove it as we solve it in the diloco PR

File: src/zeroband/train.py [hunk 16]
```diff
@@ -231,7 +246,10 @@ def train(config: Config):
             logger.info(log)
 
         if config.diloco is not None:
-            diloco.step(model)
+            with FSDP.summon_full_params(model):
```
Comment (samsja) [inline src/zeroband/train.py]: I assume this is only debugging stuff ? If yes either remove it or move it to the diloco.step with a flag to triggered it

File: scripts/simulate_multi_node_diloco.sh (hunk 3 above)
Comment (samsja) [reply scripts/simulate_multi_node_diloco.sh]: can you add to the readme or comment in the code what new env var we need to have ?

Thinking that it might be worth it to create a python wrapper around torchrun to add our new env var and potentially some custom logic. (later tho)

File: src/zeroband/testing.py [hunk 17]
```diff
@@ -0,0 +1,31 @@
+import torch
```
Comment (samsja) [inline src/zeroband/testing.py]: lets move this file to `src/zeroband/utils`

**Review by samsja — COMMENTED** [review]
I added some comments, looks good, looking forward to try !!

It could be worth to add a small tests, either torchrun one or distributed one.

Would need to test gloo over vpn asap as well

**Review by samsja — COMMENTED** [review]
you did not install pre-commit hook haha :crying_cat_face:

**Review by samsja — COMMENTED** [review]
would be nice if you add your github comments as code comments

**Comment by Jackmin801** [comment]
> you did not install pre-commit hook haha 😿

haha I did but the conflict merge broke it

**Comment by samsja** [comment]
> > you did not install pre-commit hook haha 😿
> 
> haha I did but the conflict merge broke it

ah fuck my bad

File: src/zeroband/comms.py (hunk 9 above)
Comment (Jackmin801) [reply src/zeroband/comms.py]: eh sure! The global_rank and global_world size changes though. but I think putting it all on the singleton world_info might be a good idea. lets do it

File: src/zeroband/comms.py (hunk 10 above)
Comment (Jackmin801) [reply src/zeroband/comms.py]: The global rank here is the procs rank in the global pg. The other ranks (`RANK` and `LOCAL_RANK`) are for the fsdp pg (local pg).

File: src/zeroband/comms.py (hunk 11 above)
Comment (Jackmin801) [reply src/zeroband/comms.py]: It will timeout after the TCP_STORE_TIMEOUT and raise a RuntimeError I think. This particular line shouldnt ever fail and if it does, its fatal so its fine to not retry maybe.

File: src/zeroband/comms.py (hunk 12 above)
Comment (Jackmin801) [reply src/zeroband/comms.py]: agreed!

File: src/zeroband/comms.py (hunk 13 above)
Comment (Jackmin801) [reply src/zeroband/comms.py]: ah this was before I was using the env var. it doesnt work if u have multiple diloco worker on the same machine though. will just remove it

File: src/zeroband/comms.py (hunk 14 above)
Comment (Jackmin801) [reply src/zeroband/comms.py]: ah yea makes sense. this particular line is going to be removed so not relevant here but might be relevant in the /dev/shm use in the parameter offloading. There we create a directory /dev/shm/zeroband first and put our stuff there.

File: src/zeroband/train.py (hunk 15 above)
Comment (Jackmin801) [reply src/zeroband/train.py]: yeep. happy to remove it :D

File: src/zeroband/train.py (hunk 16 above)
Comment (Jackmin801) [reply src/zeroband/train.py]: Ah I couldnt get it to work correctly without summoning full params so its not just debugging stuff.

Makes sense to move it to diloco.step though. Makes the class cleaner in the train script.

File: src/zeroband/testing.py (hunk 17 above)
Comment (Jackmin801) [reply src/zeroband/testing.py]: ye sure

File: src/zeroband/utils/__init__.py [hunk 18]
```diff
@@ -105,3 +105,33 @@ def get_model_hash(model: torch.nn.Module) -> str:
...
+        b = a.as_strided(size=(a.numel(),), stride=(1,))
+    else:
+        step_size = a.numel() // TENSOR_SIG_SAMPLE_SIZE
+        b = a.as_strided(size=(TENSOR_SIG_SAMPLE_SIZE,), stride=(step_size,))
+    element_str = "".join([f"{x:.3e}" for x in b])
+    element_hash = hashlib.md5(element_str.encode("utf-8")).hexdigest()
+    return f"{a.dtype}{a.shape}{a.stride()}<{element_hash}>"
+
+
+def get_module_signature(module: torch.nn.Module, compress: bool = True) -> str:
```
Comment (samsja) [inline src/zeroband/utils/__init__.py:115]: can you remove the get_model_hash function and change it with this one ?

File: src/zeroband/utils/world_info.py [hunk 19]
```diff
@@ -18,6 +18,15 @@ def __init__(self):
         self.local_world_size = int(os.environ["LOCAL_WORLD_SIZE"])
         self.nnodes = self.world_size // self.local_world_size
 
+        self.global_unique_id = os.environ.get("GLOBAL_UNIQUE_ID", "")
```
Comment (samsja) [inline src/zeroband/utils/world_info.py]: should we use None when the id is "" ?

File: src/zeroband/data.py [hunk 20]
```diff
@@ -78,6 +78,7 @@ def tokenize_function(data):
         train_dataset = split_dataset_by_node(tokenized_datasets, world_size=world_size, rank=rank)
 
     data_collator = collate_causal_mask(max_seq_length=seq_length, pad_id=tokenizer.pad_token_id, ignore_index=-100)
+    print(train_dataset, flush=True)
```
Comment (samsja) [inline src/zeroband/data.py]: remove print

File: src/zeroband/data.py [hunk 21]
```diff
@@ -69,7 +69,7 @@ def get_dataloader(
         ds = load_dataset("allenai/c4", "en", streaming=True)
 
         def tokenize_function(data):
-            outputs = tokenizer(data["text"], truncation=True, max_length=seq_length)
+            outputs = tokenizer(data["text"], truncation=True, max_length=seq_length, padding="max_length")
```
Comment (samsja) [inline src/zeroband/data.py]: we should not do padding here its done later in the collate funct.

File: src/zeroband/diloco.py [hunk 22]
```diff
@@ -80,25 +61,26 @@ def __init__(
...
         self._init_offloaded_optimizer(model=model)
 
-    def _init_offloaded_optimizer(self, model):
-        self.param_list_cpu = self.get_offloaded_param(model)
-        self.outer_optimizer = torch.optim.SGD(
-            self.param_list_cpu, lr=self.config.outer_lr, momentum=0.9, nesterov=True
-        )
-        self._logger.debug("offload model to cpu")
+    def _init_offloaded_optimizer(self, model: nn.Module):
+        with FSDP.summon_full_params(model):
```
Comment (samsja) [inline src/zeroband/diloco.py]: if we offload all of the weight to cpu only one worker per node should do it and then we broadcast

File: src/zeroband/diloco.py [hunk 23]
```diff
@@ -80,25 +61,26 @@ def __init__(
...
+            )
+            self._logger.debug("offload model to cpu")
 
     def sync_pseudo_gradient(self, model: nn.Module):
         """
         Sync the pseudo gradient from the local process group to the global process group
         """
         self._logger.debug("sync pseudo gradient")
+        # TODO: This assumes all params require grad, which is used by the offload
         for param_offloaded, param in zip(self.param_list_cpu, model.parameters()):
```
Comment (samsja) [inline src/zeroband/diloco.py:72]: isn't going to be a conflict here ? Since the model is sharded but the param_list_cpu is not

File: src/zeroband/diloco.py [hunk 24]
```diff
@@ -107,29 +89,51 @@ def sync_inner_model(self, model: nn.Module):
...
         self._logger.debug("sync inner model")
         for param_offloaded, param in zip(self.param_list_cpu, model.parameters()):
-            param.data.copy_(param_offloaded.data)  # todo: use copy_ here
+            param.data.copy_(param_offloaded.data)
 
     def get_offloaded_param(self, model: nn.Module) -> list[nn.Parameter]:
         """
         Offload the model parameters to cpu
         """
+        # The change here makes processes which are part of the same FSDP replica group (which are assumed to be on the same node with the same /dev/shm) use the same underlying storage for the offloaded_param.
```
Comment (samsja) [inline src/zeroband/diloco.py]: hmm okay it respond to some of my questions. Can you copy part of this comments to the docstring of the Diloco class ? it should be clear by reading the class docstring who offload to where

File: src/zeroband/diloco.py [hunk 25]
```diff
@@ -107,29 +89,51 @@ def sync_inner_model(self, model: nn.Module):
...
+        # A barrier is added to ensure that after the function completes, the parameters are all offloaded. Otherwise, the non 0 ranks might access uninitialized memory.
         offloaded_params = []
+        os.makedirs(f"/dev/shm/zeroband/{self.world_info.global_unique_id}", exist_ok=True)
 
-        for param in model.parameters():
+        for param_name, param in model.named_parameters():
             if param.requires_grad:
-                offloaded_param = param.data.detach().clone().to("cpu")
-                offloaded_param.requires_grad = True
+                storage = torch.UntypedStorage.from_file(
```
Comment (samsja) [inline src/zeroband/diloco.py]: can you confirm that this from file is actually from ram ? 
(if yes please add a comment as it could be confusing for future reader)

File: src/zeroband/diloco.py [hunk 26]
```diff
@@ -107,29 +89,51 @@ def sync_inner_model(self, model: nn.Module):
...
 
     def get_offloaded_param(self, model: nn.Module) -> list[nn.Parameter]:
         """
         Offload the model parameters to cpu
         """
+        # The change here makes processes which are part of the same FSDP replica group (which are assumed to be on the same node with the same /dev/shm) use the same underlying storage for the offloaded_param.
+        # All the processes use the same shared memory file to create a storage for each parameter. Only rank 0 will do the copy.
+        # A barrier is added to ensure that after the function completes, the parameters are all offloaded. Otherwise, the non 0 ranks might access uninitialized memory.
         offloaded_params = []
+        os.makedirs(f"/dev/shm/zeroband/{self.world_info.global_unique_id}", exist_ok=True)
```
Comment (samsja) [inline src/zeroband/diloco.py]: lets make the `/dev/shm/zeroband/` a constant

**Review by samsja — COMMENTED** [review]
I got a bit confused with the FSDP.summon_full but I think that I got it now with the shared memory

File: src/zeroband/utils/__init__.py (hunk 18 above)
Comment (Jackmin801) [reply src/zeroband/utils/__init__.py:115]: ook

File: src/zeroband/utils/world_info.py (hunk 19 above)
Comment (Jackmin801) [reply src/zeroband/utils/world_info.py]: yea might be better

File: src/zeroband/data.py (hunk 20 above)
Comment (Jackmin801) [reply src/zeroband/data.py]: ah nice catch

File: src/zeroband/diloco.py (hunk 25 above)
Comment (Jackmin801) [reply src/zeroband/diloco.py]: https://colab.research.google.com/drive/1lLwulX6nJRR1GmMPoYYOg_l9NLsRva8l?usp=sharing

yea I think so

Comment (Jackmin801) [reply src/zeroband/diloco.py]: but we have agreed to do broadcast so no sexy shm :(

File: src/zeroband/diloco.py (hunk 24 above)
Comment (Jackmin801) [reply src/zeroband/diloco.py]: reverted so no longer required

File: src/zeroband/diloco.py (hunk 23 above)
Comment (Jackmin801) [reply src/zeroband/diloco.py:72]: reverted so no longer relevant

File: src/zeroband/diloco.py (hunk 22 above)
Comment (Jackmin801) [reply src/zeroband/diloco.py]: reverted so no longer relevant

**Review by samsja — APPROVED** [review]
LFGTM !
